hibernate-core/src/main/java/org/hibernate/query/sql/internal/NativeQueryImpl.java

```diff
@@ -328,24 +328,35 @@ private void handleExplicitResultSetMapping() {
 				setTupleTransformerForResultType( resultType );
 			}
 			else {
-				checkResultType( resultType );
+				checkResultType( resultType, resultSetMapping );
 			}
 		}
 	}
 
-	private void checkResultType(Class<R> resultType) {
-		switch ( resultSetMapping.getNumberOfResultBuilders() ) {
-			case 0:
-				throw new IllegalArgumentException( "Named query exists, but did not specify a resultClass" );
-			case 1:
-				final Class<?> actualResultJavaType =
-						resultSetMapping.getResultBuilders().get( 0 ).getJavaType();
-				if ( actualResultJavaType != null && !resultType.isAssignableFrom( actualResultJavaType ) ) {
-					throw buildIncompatibleException( resultType, actualResultJavaType );
-				}
-				break;
-			default:
-				throw new IllegalArgumentException( "Cannot create TypedQuery for query with more than one return" );
+	private void checkResultType(Class<R> resultType, ResultSetMapping resultSetMapping) {
+		// resultType can be null if any of the deprecated methods were used to create the query
+		if ( resultType != null && !isResultTypeAlwaysAllowed( resultType )) {
+			switch ( resultSetMapping.getNumberOfResultBuilders() ) {
+				case 0:
+					if ( !resultSetMapping.isDynamic() ) {
+						throw new IllegalArgumentException( "Named query exists, but did not specify a resultClass" );
+					}
+					break;
+				case 1:
+					final Class<?> actualResultJavaType =
+							resultSetMapping.getResultBuilders().get( 0 ).getJavaType();
+					if ( actualResultJavaType != null && !resultType.isAssignableFrom( actualResultJavaType ) ) {
+						throw buildIncompatibleException( resultType, actualResultJavaType );
+					}
+					break;
+				default:
+					for ( ResultBuilder resultBuilder : resultSetMapping.getResultBuilders() ) {
+						final Class rbJavaType = resultBuilder.getJavaType();
+						if ( !resultType.isAssignableFrom( rbJavaType ) ) {
+							throw new IllegalArgumentException( "The result set mapping of the typed query doesn't match the declared return type" );
+						}
+					}
+			}
```

Comment:
If there are multiple `resultBuilder`s, shouldn't the `resultType` be `Object[]`?

Reply:
Not necessarily, at least that is what I thought?
If the return type is Object[], there's really not much to check since that flag covers everything.
For that reason I initially surrounded that for loop with `if ( resultType != Object[].class )` , but it should already be covered by the outer if.

Reply:
I mean this should be an error:

```java
s.createNativeQuery("select title, isbn from books", String.class)
     .addScalar("title", String.class)
     .addScalar("isbn", String.class)
     .getResultList();
```

The result type of that query has to be `Object[]`, not `String`.

Reply:
IINM, for that to work you'd have to add the entity to the native query also (`addEntity(Book.class)`) and you would still get the error (tried it)

Reply:
This query works completely perfectly:

```java
List<Object[]> resultList = 
        session.createNativeQuery("select title, isbn from book", Object[].class)
		.addScalar("title", String.class)
		.addScalar("isbn", String.class)
		.getResultList();
```

Try it!

FTR, so does this:

```java
List<Object[]> resultList = 
        session.createNativeQuery("select title, isbn from book", Object[].class)
		.getResultList();
```

but in this second case, there's nothing we can check before executing the query, of course.

Reply:
@jrenaat You only need to call `addEntity()` if the query actually returns an entity.